README.md
![Tacky](https://github.com/continuationlabs/tacky/raw/master/images/smaller.png)

<sub>Logo design by chris.ruppert@gmail.com</sub>

Server-side response caching plugin for [hapi](http://hapijs.com/)

[![Current Version](https://img.shields.io/npm/v/tacky.svg)](https://www.npmjs.org/package/tacky)
[![Build Status](https://travis-ci.org/continuationlabs/tacky.svg)](https://travis-ci.org/continuationlabs/tacky)

[![belly-button-style](https://cdn.rawgit.com/continuationlabs/belly-button/master/badge.svg)](https://github.com/continuationlabs/belly-button)

tacky adds a new handler named `cache` that can be used on any route that is a `GET` method. tacky will try to serve a value from the server cache first if present. If the value is not in the server cache, it will call `hydrate()`, reply with the result and then cache the value in the server cache for subsequent requests. tacky stores values in a hapi server cache provision. It does *not* just set the response cache headers.

## Usage

See the [API Reference](https://github.com/continuationlabs/tacky/blob/master/API.md)

### Example

_copied from examples/default.js_

```js
const Assert = require('assert');
const Http = require('http');
const Hapi = require('hapi');
const Tacky = require('tacky');

const server = new Hapi.Server();
server.connection({ port: 9001 });

server.register({ register: Tacky }, (err) => {
  Assert.ifError(err);
  server.route({
    method: 'get',
    path: '/',
    config: {
      handler: {
        cache: {
          hydrate: (request, callback) => {
            Http.get('http://www.google.com', (res) => {
              const buffers = [];
              res.on('data', (chunk) => {
                buffers.push(chunk);
              });
              res.on('end', () => {
                callback(null, buffers.join().toString());
              });
            });
          }
        }
      }
    }
  });
  server.start(() => { console.log('Server started at ' + server.info.uri); });
});
```

When the first request comes in to "/", the `hydrate` method is called. We are getting the Google home page and after 1000 milliseconds, we are calling back with the result. If you make a second request to "/", you should notice the delay isn't there and the response is almost instantaneous. The original response has been cached and sent back to the client. If you are testing with a browser, you should notice that the cache header decrements on each request. tacky will set the client cache header ("cache-control:max-age=3566, must-revalidate") based on the ttl options. The cache header ttl will be randomized so that the server isn't slammed by multiple requests at the same time. The goal is to stagger the cache header expiration across different clients.


CHANGELOG.md

## [**4.0.1**](https://github.com/continuationlabs/tacky/issues?milestone=7&state=open)
- [**#32**](https://github.com/continuationlabs/tacky/issues/32) Only hapi 13. Updated some dependencies. Closes #31.
- [**#31**](https://github.com/continuationlabs/tacky/issues/31) Drop hapi v12 from Travis

## [**4.0.0**](https://github.com/continuationlabs/tacky/issues?milestone=6&state=closed)
- [**#30**](https://github.com/continuationlabs/tacky/issues/30) Node version cleanup.
- [**#29**](https://github.com/continuationlabs/tacky/issues/29) Server.initialize
- [**#28**](https://github.com/continuationlabs/tacky/issues/28) ES6 Updates

## [**3.0.0**](https://github.com/continuationlabs/tacky/issues?milestone=3&state=closed)
- [**#26**](https://github.com/continuationlabs/tacky/issues/26) Drop semi-standard
- [**#25**](https://github.com/continuationlabs/tacky/issues/25) Hapi 10+
- [**#27**](https://github.com/continuationlabs/tacky/issues/27) Updates

## [**2.0.4**](https://github.com/continuationlabs/tacky/issues?milestone=4&state=closed)
- [**#24**](https://github.com/continuationlabs/tacky/issues/24) Testing for hapi 8-10. Lab update.

## [**2.0.1**](https://github.com/continuationlabs/tacky/issues?milestone=2&state=closed)
- [**#17**](https://github.com/continuationlabs/tacky/issues/17) Cleanup
- [**#18**](https://github.com/continuationlabs/tacky/issues/18) Misc cleanup. Closes #17.

## [**2.0.0**](https://github.com/continuationlabs/tacky/issues?milestone=1&state=closed)
- [**#13**](https://github.com/continuationlabs/tacky/issues/13) Store the state value in the cache as well
- [**#12**](https://github.com/continuationlabs/tacky/issues/12) Store `state` in Cache
- [**#11**](https://github.com/continuationlabs/tacky/issues/11) Expires
- [**#10**](https://github.com/continuationlabs/tacky/issues/10) Fixed TTL settings. Updated docs.
- [**#1**](https://github.com/continuationlabs/tacky/issues/1) Override ExpiresIn
- [**#9**](https://github.com/continuationlabs/tacky/issues/9) Changed context to be global
- [**#3**](https://github.com/continuationlabs/tacky/issues/3) Context
- [**#5**](https://github.com/continuationlabs/tacky/issues/5) No Cache
- [**#6**](https://github.com/continuationlabs/tacky/issues/6) Change State
- [**#8**](https://github.com/continuationlabs/tacky/issues/8) Undefined 2


test/index.js
'use strict';

/*eslint-disable handle-callback-err */
const Code = require('code');
const Hapi = require('hapi');
const Insync = require('insync');
const Lab = require('lab');

const Helper = require('./helper');
const Tacky = require('../lib');

const lab = exports.lab = Lab.script();
const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const internals = {
  headerRegex: /max-age=\d+,\s+must-revalidate(,\s+private)?/
};

internals.getTtl = (header) => {
  let age = header.match(/(\d+)/)[0];
  age = parseInt(age, 10);
  return age;
};

internals.checkCacheHeader = (res, expires) => {
  expect(res.headers['cache-control']).to.match(internals.headerRegex);

  const ttl = internals.getTtl(res.headers['cache-control']);
  // In seconds
  expect(ttl).to.be.about(expires, Math.floor(expires * 0.60));
};

describe('tacky', () => {
  it('throws if the handler is missing a hydrate function', (done) => {
    const server = new Hapi.Server();
    server.connection();

    server.register({
      register: Tacky
    }, (err) => {
      expect(err).to.not.exist();
      expect(() => {
        server.route({
          method: 'get',
          path: '/',
          handler: {
            cache: {}
          }
        });
      }).to.throw('hydrate must be a function.');
      done();
    });
  });

  it('throws if the method is not a "GET" method', (done) => {
    const server = new Hapi.Server();
    server.connection();

    server.register({
      register: Tacky
    }, (err) => {
      expect(err).to.not.exist();
      expect(() => {
        server.route({
          method: 'post',
          path: '/',
          handler: {
            cache: {
              hydrate () { }
            }
          }
        });
      }).to.throw('only "get" methods are supported.');
      done();
    });
  });

  it('throws if the cache does not exist', (done) => {
    const server = new Hapi.Server();
    server.connection();

    expect(() => {
      server.register({
        register: Tacky,
        options: {
          cache: 'foobar'
        }
      }, () => {});
    }).to.throw('Unknown cache foobar');
    done();
  });

  it('sets the cache-control header smartly', (done) => {
    const result = { foo: 'bar', baz: 123 };

    Helper.prepareServer({
      hydrate (request, callback) {
        callback(null, result);
      },
      expiresIn: 1000000
    }, (err, server) => {
      expect(err).to.not.exist();

      Insync.series([
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            internals.checkCacheHeader(res, 1000);
            expect(res.result).to.equal(result);
            setTimeout(next, 1000);
          });
        },
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            internals.checkCacheHeader(res, 980);
            expect(res.headers['cache-control']).to.match(internals.headerRegex);

            expect(res.result).to.equal(result);
            next();
          });
        }
      ], done);
    });
  });

  it('stores the result of hydrate in the default server memory cache', (done) => {
    const result = { foo: 'value1', bar: [1, 2, 3] };
    Helper.prepareServer({
      hydrate (request, callback) {
        callback(null, result);
      }
    }, (err, server) => {
      expect(err).to.not.exist();

      server.inject({
        url: '/'
      }, (res) => {
        internals.checkCacheHeader(res, 100);
        expect(res.result).to.equal(result);

        const cache = server._caches._default;
        expect(cache.segments['!tacky']).to.be.true();
        expect(cache.client.connection.cache['!tacky']['/']).to.exist();

        done();
      });
    });
  });

  it('stores the result of hydrate the provisioned cache', (done) => {
    const server = new Hapi.Server({
      debug: false,
      cache: {
        engine: require('hapi/node_modules/catbox-memory'),
        name: 'super-cache'
      }
    });
    server.connection();

    Insync.series([
      (next) => {
        server.register({
          register: Tacky,
          options: {
            expiresIn: 100000,
            cache: 'super-cache'
          }
        }, next);
      },
      (next) => {
        server.route({
          method: 'get',
          path: '/',
          config: {
            handler: {
              cache: {
                hydrate (request, callback) {
                  callback(null, true);
                }
              }
            }
          }
        });
        server.start(next);
      },
      (next) => {
        server.inject({
          url: '/'
        }, (res) => {
          internals.checkCacheHeader(res, 100);

          const cache = server._caches['super-cache'];
          expect(cache.segments['!tacky']).to.be.true();
          expect(cache.client.connection.cache['!tacky']['/']).to.exist();

          next();
        });
      }], done);
  });

  it('serves from the cache if present', (done) => {
    const result = 'abcdefghijk';
    let hitCount = 0;

    Helper.prepareServer({
      hydrate (request, callback) {
        hitCount++;
        callback(null, result);
      }
    }, (err, server) => {
      expect(err).to.not.exist();

      Insync.series([
        (next) => {
          server.inject({
            url: '/'
          }, next.bind(null, null));
        },
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            expect(res.result).to.equal(result);
            expect(hitCount).to.equal(1);

            next();
          });
        }
      ], done);
    });
  });

  it('always provides request state and cache information via response.plugins.tacky', (done) => {
    const result = [1, 2, 3, 4, 5];
    const state = {
      total: 1,
      data: {
        foo: 'bar',
        name: 'tacky'
      }
    };
    let hitCount = 0;

    Helper.prepareServer({
      hydrate (request, callback) {
        hitCount++;
        callback(null, result, state);
      }
    }, (err, server) => {
      expect(err).to.not.exist();

      Insync.series([
        (next) => {
          server.ext('onPreResponse', (request, reply) => {
            const tacky = request.response.plugins.tacky;
            expect(tacky.state).to.equal(state);
            expect(tacky.cache).to.exist();
            expect(tacky.cache).to.have.length(3);
            reply.continue();
          });

          next();
        },
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            expect(res.result).to.equal(result);
            next();
          });
        },
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            expect(res.result).to.equal(result);
            next();
          });
        }
      ], (err) => {
        expect(hitCount).to.equal(1);
        done();
      });
    });
  });

  it('will report cache errors when getting cache values', (done) => {
    const result = [{ foo: 'bar' }, { foo: 'baz' }, { foo: 'zip' }];

    Helper.prepareServer({
      hydrate (request, callback) {
        callback(null, result);
      }
    }, (err, server) => {
      expect(err).to.not.exist();
      Insync.series([
        (next) => {
          server.once('request', (request, event) => {
            expect(event.tags).to.equal(['cache', 'error']);
            expect(event.data.message).to.equal('Error looking up / in the cache');
          });
          next();
        },
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            internals.checkCacheHeader(res, 100);
            expect(res.result).to.equal(result);
            next();
          });
        }
      ], done);
    });
  });

  it('will report cache errors when setting cache values', (done) => {
    const result = [{ foo: 'bar' }, { foo: 'baz' }, { foo: 'zip' }];

    Helper.prepareServer({
      hydrate (request, callback) {
        callback(null, result);
      }
    }, (err, server) => {
      expect(err).to.not.exist();
      Insync.series([
        (next) => {
          server.once('request', (request, event) => {
            expect(event.tags).to.equal(['cache', 'error']);
            expect(event.data.message).to.equal('Error setting cache for /');
          });
          next();
        },
        (next) => {
          const Policy = require('hapi/node_modules/catbox/lib/policy');
          const set = Policy.prototype.set;
          Policy.prototype.set = (id, value, options, callback) => {
            Policy.prototype.set = set;

            expect(id).to.equal('/');
            expect(value).to.equal({
              result: result,
              state: null
            });
            callback(new Error('mock error for testing'));
          };

          server.inject({
            url: '/'
          }, (res) => {
            internals.checkCacheHeader(res, 100);
            expect(res.result).to.equal(result);
            next();
          });
        }
      ], done);
    });
  });

  it('will reply with an error if hydrate fails', (done) => {
    Helper.prepareServer({
      hydrate (request, callback) {
        callback(new Error('mock error for testing'));
      }
    }, (err, server) => {
      expect(err).to.not.exist();
      server.inject({
        url: '/'
      }, (res) => {
        expect(res.result).to.equal({
          statusCode: 500,
          error: 'Internal Server Error',
          message: 'An internal server error occurred'
        });
        done();
      });
    });
  });

  it('will pass the context option to the hydrate and generateKey function', (done) => {
    const server = new Hapi.Server({ debug: false });
    const context = {
      foo: 'bar',
      baz: true
    };
    server.connection();

    Insync.series([
      (next) => {
        server.register({
          register: Tacky,
          options: { expiresIn: 100000, bind: context }
        }, next);
      },
      (next) => {
        server.route({
          method: 'get',
          path: '/',
          config: {
            handler: {
              cache: {
                hydrate (request, callback) {
                  expect(this).to.equal(context);
                  callback(null, true);
                },
                generateKey (request) {
                  expect(this).to.equal(context);
                  return request.raw.req.url;
                },
                privacy: 'private'
              }
            }
          }
        });
        server.start(next);
      },
      (next) => {
        server.inject({
          url: '/'
        }, (res) => {
          internals.checkCacheHeader(res, 100);
          next();
        });
      }], done);
  });

  it('will not cache values if the generateKey is null or undefined', (done) => {
    const result = { foo: 'bar', baz: 123 };
    Helper.prepareServer({
      hydrate (request, callback) {
        callback(null, result);
      },
      generateKey () {
        return undefined;
      }
    }, (err, server) => {
      expect(err).to.not.exist();

      server.ext('onPreResponse', (request, reply) => {
        expect(request.response.plugins.tacky).to.equal({
          cache: null,
          state: null
        });
        reply.continue();
      });

      Insync.series([
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            expect(res.statusCode).to.equal(200);
            expect(res.result).to.equal(result);

            const cache = server._caches._default;
            expect(cache.segments['!tacky']).to.be.true();
            expect(cache.client.connection.cache).to.have.length(0);
            done();
          });
        }
      ], done);
    });
  });

  it('will use the expiresIn route option instead of the Policy default', (done) => {
    const result = [{ foo: 'bar' }, { foo: 'baz' }, { foo: 'zip' }];

    Helper.prepareServer({
      hydrate (request, callback) {
        callback(null, 123456);
      }
    }, (err, server) => {
      Insync.series([
        (next) => {
          server.route({
            method: 'get',
            path: '/cache',
            config: {
              handler: {
                cache: {
                  hydrate (request, callback) {
                    return callback(null, result);
                  },
                  expiresIn: 1000000
                }
              }
            }
          });
          next();
        },
        (next) => {
          server.inject({
            url: '/cache'
          }, (res) => {
            internals.checkCacheHeader(res, 1000);
            expect(res.result).to.equal(result);

            const cache = server._caches._default;
            expect(cache.segments['!tacky']).to.be.true();
            expect(cache.client.connection.cache['!tacky']['/cache'].ttl).to.equal(1000000);
            expect(cache.client.connection.cache['!tacky']['/cache'].item).to.equal(JSON.stringify({ result: result, state: null }));

            next();
          });
        },
        (next) => {
          server.inject({
            url: '/'
          }, (res) => {
            internals.checkCacheHeader(res, 100);
            expect(res.result).to.equal(123456);

            const cache = server._caches._default;
            expect(cache.client.connection.cache['!tacky']['/'].ttl).to.equal(100000);
            expect(cache.client.connection.cache['!tacky']['/'].item).to.equal(JSON.stringify({ result: 123456, state: null }));

            next();
          });
        }], done);
    });
  });

  describe('generateKey()', () => {
    it('can be used to override the default generateKey function', (done) => {
      const result = { foo: 'bar', baz: 123 };
      Helper.prepareServer({
        hydrate (request, callback) {
          callback(null, result);
        },
        generateKey (request) {
          return 12345 + request.state.foo;
        }
      }, (err, server) => {
        expect(err).to.not.exist();
        Insync.series([
          (next) => {
            server.inject({
              url: '/',
              headers: {
                'Cookie': 'foo=bar'
              }
            }, (res) => {
              expect(res.headers['cache-control']).to.match(internals.headerRegex);

              const cache = server._caches._default;
              expect(cache.segments['!tacky']).to.be.true();
              expect(cache.client.connection.cache['!tacky']['12345bar']).to.exist();

              next();
            });
          }
        ], done);
      });
    });

    it('will throw an error if the resultant cache key is not a string', (done) => {
      const result = { foo: 'bar', baz: 123 };

      Helper.prepareServer({
        hydrate (request, callback) {
          callback(null, result);
        },
        generateKey (request) {
          return false;
        }
      }, (err, server) => {
        expect(err).to.not.exist();
        Insync.series([
          (next) => {
            server.on('request-error', (request, err) => {
              expect(err).to.be.an.instanceOf(Error);
              expect(err.message).to.equal('Uncaught error: generateKey must return a string.');
            });
            next();
          },
          (next) => {
            server.inject({
              url: '/'
            }, next.bind(null, null));
          }
        ], done);
      });
    });
  });
});


test/helper.js
'use strict';

const Hapi = require('hapi');
const Insync = require('insync');
const Tacky = require('../lib');

module.exports.prepareServer = (config, callback) => {
  /*eslint-disable */
  config.expiresIn = config.expiresIn || 100000;
  /*eslint-enable */
  const server = new Hapi.Server({ debug: false });
  server.connection();

  Insync.series([
    (next) => {
      server.register({
        register: Tacky,
        options: { expiresIn: config.expiresIn }
      }, next);
    },
    (next) => {
      server.route({
        method: 'get',
        path: '/',
        config: {
          handler: {
            cache: {
              hydrate: config.hydrate,
              generateKey: config.generateKey
            }
          }
        }
      });
      next();
    },
    (next) => {
      server.initialize(next);
    }
  ], function () {
    callback(null, server);
  });
};


examples/state.js
'use strict';

const Assert = require('assert');
const Http = require('http');
const Hapi = require('hapi');
const Reach = require('reach');
const Tacky = require('../lib');

const server = new Hapi.Server();
server.connection({ port: 9001 });

server.register({ register: Tacky }, (err) => {
  Assert.ifError(err);
  server.route({
    method: 'get',
    path: '/',
    config: {
      handler: {
        cache: {
          hydrate: (request, callback) => {
            Http.get('http://www.google.com', (res) => {
              const buffers = [];
              res.on('data', (chunk) => { buffers.push(chunk); });
              res.on('end', () => {
                callback(null, buffers.join().toString(), {
                  statusCode: 202
                });
              });
            });
          }
        }
      }
    }
  });
  server.start(() => { console.log('Server started at ' + server.info.uri); });
  server.ext('onPreResponse', (request, reply) => {
    const response = request.response;

    if (response.isBoom) {
      return reply.continue();
    }

    const state = Reach(response, 'plugins.tacky.state');

    if (state) {
      response.statusCode = state.statusCode;
    }
    reply.continue();
  });
});


examples/default.js
'use strict';

const Assert = require('assert');
const Http = require('http');
const Hapi = require('hapi');
const Tacky = require('../lib');

const server = new Hapi.Server();
server.connection({ port: 9001 });

server.register({ register: Tacky }, (err) => {
  Assert.ifError(err);
  server.route({
    method: 'get',
    path: '/',
    config: {
      handler: {
        cache: {
          hydrate: (request, callback) => {
            Http.get('http://www.google.com', (res) => {
              const buffers = [];
              res.on('data', (chunk) => {
                buffers.push(chunk);
              });
              res.on('end', () => {
                callback(null, buffers.join().toString());
              });
            });
          }
        }
      }
    }
  });
  server.start(() => { console.log('Server started at ' + server.info.uri); });
});


lib/index.js
'use strict';

const Assert = require('assert');
const Insync = require('insync');
const Reach = require('reach');
const Merge = require('lodash.merge');

const defaults = {
  expiresIn: 3600000,
  privacy: 'default'
};
const pkg = require('../package.json');
const NAME = pkg.name;

const internals = {};

module.exports.register = function (server, options, next) {
  internals.settings = Merge({}, defaults, options);
  internals.cache = server.cache({
    expiresIn: internals.settings.expiresIn,
    cache: internals.settings.cache
  });

  server.ext('onPreResponse', internals.extensionPoint);
  server.handler('cache', internals.handler.bind(options.bind));
  next();
};

module.exports.register.attributes = { pkg };

internals.handler = function (route, options) {
  Assert.equal(typeof options.hydrate, 'function', 'hydrate must be a function.');
  Assert.equal(route.method, 'get', 'only "get" methods are supported.');

  // This is to prevent confusion about what cach headers to use when sending the response
  // any cache related settings should be under `options` and not under route.settings.cache
  delete route.settings.cache;

  const settings = Merge({}, {
    generateKey (request) {
      return request.raw.req.url;
    }
  }, options);

  const self = this;
  return function (request, reply) {
    const cacheKey = settings.generateKey.call(self, request);
    const privacy = options.privacy || internals.settings.privacy;
    // If the ttl hasn't been overwritten by a route option, use the default cache policy value which is keyed as 0 via CatBox
    // this is what we use thoughtout the code to call out to Catbox
    const ttl = options.expiresIn || 0;
    // This will be passed out in request.state.tacky so it can be used to set cache-headers
    const maxAge = options.expiresIn || internals.settings.expiresIn;

    // Waterfall functions
    const hydrate = settings.hydrate.bind(self, request);
    const done = function (err, data) {
      if (err) {
        return reply(err);
      }

      const response = reply(data.result);
      response.plugins[NAME] = {
        cache: data.cache,
        state: data.state
      };
    };
    const afterHydrate = function (cache) {
      return function (result /*, [state], [next]*/) {
        let next;
        let state;
        if (arguments.length === 3) {
          state = arguments[1];
          next = arguments[2];
        } else {
          state = null;
          next = arguments[1];
        }

        next(null, { result, state, cache });
      };
    };
    const checkCache = function (next) {
      internals.cache.get(cacheKey, function (err, value, cached) {
        if (err) {
          request.log(['cache', 'error'], {
            message: `Error looking up ${cacheKey} in the cache`,
            error: err
          });
        }

        // If the value is in the cache, short-circuit the waterfall and
        // call out to the end. This is the "documented" way to short-circuit
        // an async waterfall (https://github.com/caolan/async/pull/85#issuecomment-13072390).
        if (cached) {
          return done(null, {
            result: value.result,
            cache: {
              ttl: cached.ttl,
              maxAge: cached.ttl,
              privacy: privacy
            },
            state: value.state
          });
        }
        next(null);
      });
    };

    const tasks = [];
    /* eslint-disable*/
    if (cacheKey == null) {
    /* eslint-enable*/
      tasks.push(hydrate, afterHydrate(null));
    } else {
      tasks.push(checkCache, hydrate);
      tasks.push(afterHydrate({ ttl, maxAge, privacy }));
      tasks.push(function (arg, next) {
        const tail = request.tail('cache tail');
        const value = {
          result: arg.result,
          state: arg.state
        };
        internals.cache.set(cacheKey, value, arg.cache.ttl, function (cacheErr) {
          if (cacheErr) {
            request.log(['cache', 'error'], {
              message: `Error setting cache for ${cacheKey}`,
              error: cacheErr
            });
          }
          tail();
        });
        next(null, arg);
      });
    }

    Insync.waterfall(tasks, done);
  };
};

internals.getCacheString = function (ttl, privacy) {
  const age = Math.floor(ttl / 1000);
  let header = `max-age=${age}, must-revalidate`;

  if (privacy !== 'default') {
    header += ', ' + privacy;
  }
  return header;
};

internals.extensionPoint = function (request, reply) {
  const response = request.response;

  if (response.isBoom) {
    return reply.continue();
  }

  const cache = Reach(response, `plugins.${NAME}.cache`);
  if (cache) {
    // Random number between 60% of the ttl and the TTL.
    const age = internals.between(cache.maxAge * 0.60, cache.maxAge);

    response.header('cache-control', internals.getCacheString(age, cache.privacy));
  }
  reply.continue();
};

internals.between = function (min, max) {
  min = Math.floor(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1) + min);
};
